### ISSUE TITLE
New Feature:  Allow KernelFunction method implementations to include metadata 

### ISSUE THREAD

**Issue by lavinir** [microsoft/semantic-kernel#7071]
Currently, there isn't a way for a Kernel Function to return additional data that may be useful to the app flow, but that is not strictly part of the function Result object. 
For example, a function that answers queries about a document should return the relevant information as part of the function result, but it would be useful to have something like citations added as additional metadata that can be used. 
Sometimes the metadata may also be used to determine further instructions to the model when sending back the information etc.
If perhaps a dedicated object type can be used when defining Kernel Functions that can contain both the data response and additional metadata which can then be appropriately populated in the **FunctionResult** class, that would be very helpful.

**Comment by dmytrostruk** [comment]
> For example, a function that answers queries about a document should return the relevant information as part of the function result, but it would be useful to have something like citations added as additional metadata that can be used.
Sometimes the metadata may also be used to determine further instructions to the model when sending back the information etc.

@lavinir When you define functions, you decide which result it should return and what information to share with the model. If you need additional metadata returned from your function, you can define your custom type, return it from function, it will be serialized and shared with the model.

> If perhaps a dedicated object type can be used when defining Kernel Functions that can contain both the data response and additional metadata which can then be appropriately populated in the FunctionResult class, that would be very helpful.

Today you can return `FunctionResult` directly from your function, if that will cover your scenario. The only limitation is that `FunctionResult` constructor should accept `KernelFunction` as a reference, so in your function you will need `Kernel` as input parameter to be able to fetch current function instance from Kernel:

```csharp
public async Task TestAsync()
{
    var kernel = Kernel.CreateBuilder().Build();

    kernel.ImportPluginFromType<MyPlugin>();

    var result = await kernel.InvokeAsync("MyPlugin", "GetResult");

    Console.WriteLine(result.ToString()); // test result
    Console.WriteLine(result.Metadata["key"]); // value
}

public sealed class MyPlugin
{
    [KernelFunction]
    public FunctionResult GetResult(Kernel kernel)
    {
        var thisFunction = kernel.Plugins[nameof(MyPlugin)][nameof(GetResult)];

        var result = "test result";
        var metadata = new Dictionary<string, object?> { ["key"] = "value" };

        return new FunctionResult(thisFunction, result, metadata: metadata);
    }
}
```

But I still think that using your custom type should work better based on use-cases you shared. Please let me know if it's helpful. Thanks!

**Comment by dmytrostruk** [comment]
@lavinir Based on my previous comment, I'm closing this issue for now. Feel free to leave your feedback under this issue or create a new one if needed. Thanks!

**Comment by dmytrostruk** [comment]
> Likewise for FunctionResult, correct me if I'm wrong but it seems if that type is returned it is also serialized (including the metadata) before being sent to the model.

@lavinir No, when you return `FunctionResult` from your function as shared in my previous code example, this type is handled in special way. Unlike other custom type, we do not serialize it, we just take a `Result` property value from it and share it with the model, which means that metadata is not included. I just verified this behavior, and it works as I described.
